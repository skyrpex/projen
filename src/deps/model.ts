export interface DepsManifest {
  /**
   * All dependencies of this module.
   */
  readonly dependencies: Dependency[];
}

export interface Dependency {
  /**
   * The package manager name of the dependency (e.g. `leftpad` for npm).
   *
   * NOTE: For package managers that use complex coordinates (like Maven), we
   * will codify it into a string somehow.
   */
  readonly name: string;

  /**
   * Which type of dependency this is (runtime, build-time, etc).
   */
  readonly type: DependencyType;

  /**
   * Semantic version version requirement.
   *
   * If the dependency is part of a mono-repo, use `monorepo`.
   *
   * @default - requirement is managed by the package manager (e.g. npm/yarn).
   */
  readonly version?: string;
}

export enum DependencyType {
  /**
   * The dependency is required for the program/library during runtime.
   */
  RUNTIME = 'runtime',

  /**
   * The dependency is required at runtime but expected to be installed by the
   * consumer.
   */
  PEER = 'peer',

  /**
   * The dependency is bundled and shipped with the module, so consumers are not
   * required to install it.
   */
  BUNDLED = 'bundled',

  /**
   * The dependency is required to run the `build` task.
   */
  BUILD = 'build',

  /**
   * The dependency is required to run the `test` task.
   */
  TEST = 'test',

  /**
   * The dependency is required for development (e.g. IDE plugins).
   */
  DEVENV = 'devenv',
}
