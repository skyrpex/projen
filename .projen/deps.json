{
  "dependencies": [
    {
      "name": "@types/fs-extra",
      "version": "^8",
      "type": "build"
    },
    {
      "name": "@types/glob",
      "type": "build"
    },
    {
      "name": "@types/inquirer",
      "type": "build"
    },
    {
      "name": "@types/jest",
      "type": "build"
    },
    {
      "name": "@types/node",
      "version": "^10.17.0",
      "type": "build"
    },
    {
      "name": "@types/semver",
      "type": "build"
    },
    {
      "name": "@types/yargs",
      "type": "build"
    },
    {
      "name": "@typescript-eslint/eslint-plugin",
      "version": "^4.3.0",
      "type": "build"
    },
    {
      "name": "@typescript-eslint/parser",
      "version": "^4.3.0",
      "type": "build"
    },
    {
      "name": "eslint",
      "type": "build"
    },
    {
      "name": "eslint-import-resolver-node",
      "type": "build"
    },
    {
      "name": "eslint-import-resolver-typescript",
      "type": "build"
    },
    {
      "name": "eslint-plugin-import",
      "type": "build"
    },
    {
      "name": "jest",
      "type": "build"
    },
    {
      "name": "jest-junit",
      "version": "^12",
      "type": "build"
    },
    {
      "name": "jsii",
      "type": "build"
    },
    {
      "name": "jsii-diff",
      "type": "build"
    },
    {
      "name": "jsii-docgen",
      "type": "build"
    },
    {
      "name": "jsii-pacmak",
      "type": "build"
    },
    {
      "name": "jsii-release",
      "type": "build"
    },
    {
      "name": "json-schema",
      "type": "build"
    },
    {
      "name": "markmac",
      "type": "build"
    },
    {
      "name": "standard-version",
      "version": "^9.0.0",
      "type": "build"
    },
    {
      "name": "ts-jest",
      "type": "build"
    },
    {
      "name": "typescript",
      "version": "^3.9.5",
      "type": "build"
    },
    {
      "name": "@iarna/toml",
      "type": "bundled"
    },
    {
      "name": "chalk",
      "type": "bundled"
    },
    {
      "name": "decamelize",
      "type": "bundled"
    },
    {
      "name": "fs-extra",
      "type": "bundled"
    },
    {
      "name": "glob",
      "version": "^7",
      "type": "bundled"
    },
    {
      "name": "inquirer",
      "type": "bundled"
    },
    {
      "name": "semver",
      "type": "bundled"
    },
    {
      "name": "yaml",
      "type": "bundled"
    },
    {
      "name": "yargs",
      "type": "bundled"
    }
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"npx projen\"."
}