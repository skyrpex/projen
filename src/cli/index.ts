import * as yargs from 'yargs';
import { synth } from './synth';
import { discoverTaskCommands } from './tasks';

async function main() {
  const ya = yargs;
  ya.commandDir('cmds');

  discoverTaskCommands(ya);

  ya.recommendCommands();
  ya.wrap(yargs.terminalWidth());
  ya.option('post', { type: 'boolean', default: true, desc: 'Run post-synthesis steps such as installing dependencies. Use --no-post to skip' });
  ya.options('debug', { type: 'boolean', default: false, desc: 'Debug logs' });
  ya.help();

  const args = ya.argv;

  if (args.debug) {
    process.env.DEBUG = 'true';
  }

  // no command means just require .projenrc.js
  if (args._.length === 0) {
    process.env.PROJEN_DISABLE_POST = (!args.post).toString();
    await synth();
  }
}


main().catch(e => {
  console.error(e.stack);
  process.exit(1);
});
