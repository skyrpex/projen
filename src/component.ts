import { Project } from './project';

/**
 * Represents a project component.
 */
export class Component {
  constructor(public readonly project: Project) {
    project._addComponent(this);
  }

  public preSynthesize() { }

  /**
   * Synthesizes files to the project output directory.
   */
  public synthesize() {}

  /**
   * Called after synthesis. Order is *not* guaranteed.
   */
  public postSynthesize() {}
}