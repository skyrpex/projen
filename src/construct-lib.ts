import { JsiiProject, JsiiProjectOptions } from './jsii-project';

export interface ConstructLibraryOptions extends JsiiProjectOptions {
  /**
   * Libraries will be picked up by the construct catalog when they are
   * published to npm as jsii modules and will be published under:
   *
   *     https://awscdk.io/packages/[@SCOPE/]PACKAGE@VERSION
   *
   * The catalog will also post a tweet to https://twitter.com/awscdkio with the
   * package name, description and the above link. You can disable these tweets
   * through `{ announce: false }`.
   *
   * You can also add a Twitter handle through `{ twitter: 'xx' }` which will be
   * mentioned in the tweet.
   *
   * @see https://github.com/construct-catalog/catalog
   *
   * @default - new version will be announced
   */
  readonly catalog?: Catalog;
}

export interface Catalog {
  /**
   * Twitter account to @mention in announcement tweet.
   */
  readonly twitter?: string;

  /**
   * Should we announce new versions?
   * @default true
   */
  readonly announce?: boolean;
}

/**
 * A multi-language library for CDK constructs.
 */
export abstract class ConstructLibrary extends JsiiProject {
  constructor(options: ConstructLibraryOptions) {
    super(options);

    this.addKeywords('cdk'); // publish to the catalog
    if (options.catalog) {
      this.npmPackage.addField('awscdkio', {
        twitter: options.catalog.twitter,
        announce: options.catalog.announce,
      });
    }
  }
}
