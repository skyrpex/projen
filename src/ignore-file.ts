import { FileBase, IResolver } from './file';
import { Project } from './project';


export class IgnoreFile extends FileBase {
  private readonly _excludes = new Set<string>();
  private readonly _includes = new Set<string>();

  constructor(project: Project, filePath: string) {
    super(project, filePath, { editGitignore: filePath !== '.gitignore' });
  }

  /**
   * Add ignore patterns. Files that match this pattern will be ignored. If the
   * pattern starts with a negation mark `!`, files that match will _not_ be
   * ignored.
   *
   * Comment lines (start with `#`) are ignored.
   *
   * @param patterns Ignore patterns.
   */
  public addPatterns(...patterns: string[]) {
    for (const pattern of patterns) {
      // skip comments
      if (pattern.startsWith('#')) {
        continue;
      }

      if (pattern.startsWith('!')) {
        this._includes.add(pattern);
      } else {
        this._excludes.add(pattern);
      }
    }
  }

  /**
   * Removes patterns previously added from the ignore file.
   *
   * If `addPattern()` is called after this, the pattern will be added again.
   *
   * @param patterns patters to remove.
   */
  public removePatterns(...patterns: string[]) {
    for (const p of patterns) {
      this._excludes.delete(p);
      this._includes.delete(p);
    }
  }

  /**
   * Ignore the files that match these patterns.
   * @param patterns The patterns to match.
   */
  public exclude(...patterns: string[]) {
    return this.addPatterns(...patterns);
  }

  /**
   * Always include the specified file patterns.
   * @param patterns Patterns to include in git commits.
   */
  public include(...patterns: string[]) {
    for (let pattern of patterns) {
      if (!pattern.startsWith('!')) {
        pattern = '!' + pattern;
      }

      this.addPatterns(pattern);
    }
  }

  protected synthesizeContent(resolver: IResolver): string | undefined {
    const lines = [`# ${FileBase.PROJEN_MARKER}`];

    for (const line of Array.from(this._excludes).sort()) {
      // if this line exists in the inclusion list, then skip it
      if (this._includes.has(`!${line}`)) {
        continue;
      }

      lines.push(line);
    }

    for (const line of Array.from(this._includes).sort()) {
      lines.push(line);
    }

    lines.push('\n');

    return resolver.resolve(lines).join('\n');
  }
}
