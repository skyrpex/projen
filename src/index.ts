export * from './awscdk-app-ts';
export * from './awscdk-construct';
export * from './cdk8s-construct';
export * from './component';
export * from './construct-lib';
export * from './docker-compose';
export * from './eslint';
export * from './file';
export * from './gitpod';
export * from './ignore-file';
export * from './jest';
export * from './jsii-project';
export * from './json';
export * from './license';
export * from './makefile';
export * from './node-project';
export * from './project';
export * from './readme';
export * from './sample-file';
export * from './semver';
export * from './textfile';
export * from './toml';
export * from './typescript';
export * from './version';
export * from './yaml';

// export submobules
export * as tasks from './tasks';
export * as web from './web';
export * as github from './github';
export * as vscode from './vscode';
export * as deps from './deps';