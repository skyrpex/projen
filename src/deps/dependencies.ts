import * as path from 'path';
import { PROJEN_DIR } from '../common';
import { Component } from '../component';
import { JsonFile } from '../json';
import * as logging from '../logging';
import { Project } from '../project';
import { Dependency, DependencyType, DepsManifest } from './model';

/**
 * The `Dependencies` component is responsible to track the list of dependencies
 * a project has, and then used by project types as the model for rendering
 * project-specific dependency manifests such as the dependencies section
 * `package.json` files.
 *
 * To add a dependency you can use a project-type specific API such as
 * `nodeProject.addDeps()` or use the generic API of `project.deps`:
 */
export class Dependencies extends Component {
  /**
   * The project-relative path of the deps manifest file.
   */
  public static readonly MANIFEST_FILE = path.join(PROJEN_DIR, 'deps.json');

  private readonly _deps = new Array<Dependency>();

  /**
   * Adds a dependencies component to the project.
   * @param project The parent project
   */
  constructor(project: Project) {
    super(project);

    // this is not really required at the moment, but actually quite useful as a
    // checked-in source of truth for dependencies and will potentially be
    // valuable in the future for CLI tools.
    new JsonFile(project, Dependencies.MANIFEST_FILE, {
      marker: true,
      omitEmpty: true,
      obj: () => this.toJson(),
    });
  }

  /**
   * A copy of all dependencies recorded for this project.
   *
   * The list is sorted by type->name->version
   */
  public get all(): Dependency[] {
    return [...this._deps].sort(compareDeps);
  }

  /**
   * Returns a dependency by name.
   *
   * Fails if there is no dependency defined by that name or if `type` is not
   * provided and there is more then one dependency type for this dependency.
   *
   * @param name The name of the dependency
   * @param type The dependency type. If this dependency is defined only for a
   * single type, this argument can be omitted.
   *
   * @returns a copy (cannot be modified)
   */
  public getDependency(name: string, type?: DependencyType): Dependency {
    const idx = this.tryGetDependencyIndex(name, type);
    if (idx === -1) {
      const msg = type
        ? `there is no ${type} dependency defined on "${name}"`
        : `there is no dependency defined on "${name}"`;

      throw new Error(msg);
    }

    return {
      ...this._deps[idx],
    };
  }

  /**
   * Adds a dependency to this project.
   * @param spec The dependency spec in the format `MODULE[@VERSION]` where
   * `MODULE` is the package-manager-specific module name and `VERSION` is an
   * optional semantic version requirement (e.g. `^3.4.0`).
   * @param type The type of the dependency.
   */
  public addDependency(spec: string, type: DependencyType): Dependency {
    logging.debug(`${type}-dep ${spec}`);

    const scope = spec.startsWith('@');
    if (scope) {
      spec = spec.substr(1);
    }

    const [module, version] = spec.split('@');
    const name = scope ? `@${module}` : module;

    const dep: Dependency = version
      ? { name, version, type }
      : { name, type };

    this._deps.push(dep);
    return dep;
  }

  /**
   * Removes a dependency.
   * @param name The name of the module to remove (without the version)
   * @param type The dependency type. This is only required if there the
   * dependency is defined for multiple types.
   */
  public removeDependency(name: string, type?: DependencyType) {
    const removeIndex = this.tryGetDependencyIndex(name, type);
    if (removeIndex === -1) {
      return;
    }

    this._deps.splice(removeIndex, 1);
  }

  private tryGetDependencyIndex(name: string, type?: DependencyType): number {
    const deps = this._deps.filter(d => d.name === name);
    if (deps.length === 0) {
      return -1; // not found
    }

    if (!type) {
      if (deps.length > 1) {
        throw new Error(`"${name}" is defined for multiple dependency types: ${deps.map(d => d.type).join(',')}. Please specify dependency type`);
      }

      type = deps[0].type;
    }

    return this._deps.findIndex(dep => dep.name === name && dep.type === type);
  }

  private toJson(): DepsManifest {
    return {
      dependencies: this._deps.sort(compareDeps),
    };
  }
}

function compareDeps(d1: Dependency, d2: Dependency) {
  return specOf(d1).localeCompare(specOf(d2));

  function specOf(dep: Dependency) {
    let spec = dep.type + ':' + dep.name;
    if (dep.version) {
      spec += '@' + dep.version;
    }
    return spec;
  }
}

