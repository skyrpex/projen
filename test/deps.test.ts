import { join } from 'path';
import { readFileSync } from 'fs-extra';
import { Project } from '../src';
import { Dependencies, DependencyType } from '../src/deps';
import { TestProject } from './util';

test('no dependencieds, empty manifest', () => {
  // GIVEN
  const p = new TestProject();

  // THEN
  expect(depsManifest(p)).toStrictEqual({
    '//': '~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".',
  });
});

test('minimal case', () => {
  // GIVEN
  const p = new TestProject();

  // WHEN
  const dep = p.deps.addDependency('my-first-dep', DependencyType.RUNTIME);

  // THEN
  expect(dep.name).toBe('my-first-dep');
  expect(dep.version).toBeUndefined();
  expect(dep.type).toBe(DependencyType.RUNTIME);
  expect(depsManifest(p)).toMatchSnapshot();
});

test('with version requirement', () => {
  // GIVEN
  const p = new TestProject();

  // WHEN
  const dep = p.deps.addDependency('depy@^7', DependencyType.PEER);

  // THEN
  expect(dep.name).toBe('depy');
  expect(dep.version).toBe('^7');
  expect(dep.type).toBe(DependencyType.PEER);

  expect(depsManifest(p)).toMatchSnapshot();
});

test('deps.all returns all the dependencies', () => {
  // GIVEN
  const p = new TestProject();

  // WHEN
  p.deps.addDependency('dep1', DependencyType.DEVENV);
  p.deps.addDependency('dep2', DependencyType.RUNTIME);
  p.deps.addDependency('dep3', DependencyType.PEER);

  // THEN
  expect(p.deps.all).toMatchSnapshot();
});

describe('removeDependency()', () => {
  test('can be used to remove a dependency', () => {
    // GIVEN
    const p = new TestProject();
    p.deps.addDependency('mydep', DependencyType.RUNTIME);

    // WHEN
    p.deps.removeDependency('mydep');

    // THEN
    expect(p.deps.all).toStrictEqual([]);
  });

  test('dep is defined for multiple types (e.g. dev + runtime)', () => {
    // GIVEN
    const p = new TestProject();
    p.deps.addDependency('mydep', DependencyType.RUNTIME);
    p.deps.addDependency('mydep', DependencyType.BUILD);

    // WHEN
    p.deps.removeDependency('mydep', DependencyType.BUILD);

    // THEN
    expect(p.deps.all).toStrictEqual([
      {
        name: 'mydep',
        type: 'runtime',
      },
    ]);

  });

  test('fails if type is not provided and there are more then one', () => {
    // GIVEN
    const p = new TestProject();

    // WHEN
    p.deps.addDependency('foo', DependencyType.BUILD);
    p.deps.addDependency('foo', DependencyType.RUNTIME);

    // THEN
    expect(() => p.deps.removeDependency('foo')).toThrow(/"foo" is defined for multiple dependency types\: build\,runtime/);
  });

  test('no-op if the dependency is not defined', () => {
    // GIVEN
    const p = new TestProject();
    p.deps.addDependency('hey', DependencyType.RUNTIME);

    // WHEN
    p.deps.removeDependency('bam');

    // THEN
    expect(p.deps.all).toStrictEqual([
      {
        name: 'hey',
        type: 'runtime',
      },
    ]);
  });
});

describe('getDependency()', () => {

  test('returns a single dependency', () => {
    // GIVEN
    const p = new TestProject();

    // WHEN
    p.deps.addDependency('bar@^1.1.1', DependencyType.RUNTIME);

    // THEN
    expect(p.deps.getDependency('bar')).toStrictEqual({
      name: 'bar',
      version: '^1.1.1',
      type: DependencyType.RUNTIME,
    });
  });

  test('matches type', () => {
    // GIVEN
    const p = new TestProject();
    p.deps.addDependency('boo', DependencyType.BUILD);
    p.deps.addDependency('zar', DependencyType.RUNTIME);
    p.deps.addDependency('hey@^1.0.0', DependencyType.RUNTIME);

    // WHEN
    p.deps.addDependency('bar@^1.1.1', DependencyType.RUNTIME);

    // THEN
    expect(p.deps.getDependency('bar', DependencyType.RUNTIME)).toStrictEqual({
      name: 'bar',
      version: '^1.1.1',
      type: DependencyType.RUNTIME,
    });
  });

  test('matches type (multiple)', () => {
    // GIVEN
    const p = new TestProject();

    // WHEN
    p.deps.addDependency('bar@^1.2.1', DependencyType.RUNTIME);
    p.deps.addDependency('bar@^1.1.1', DependencyType.BUILD);

    // THEN
    expect(p.deps.getDependency('bar', DependencyType.RUNTIME)).toStrictEqual({
      name: 'bar',
      version: '^1.2.1',
      type: DependencyType.RUNTIME,
    });
  });

  test('fails if there is no dependency by that name', () => {
    // GIVEN
    const p = new TestProject();
    p.deps.addDependency('bar@1.1.1', DependencyType.RUNTIME);

    // THEN
    expect(() => p.deps.getDependency('test')).toThrow(/there is no dependency defined on \"test\"/);
  });

  test('fails if there is more then one type and type is not provided', () => {
    // GIVEN
    const p = new TestProject();

    // WHEN
    p.deps.addDependency('zoo', DependencyType.RUNTIME);
    p.deps.addDependency('zoo', DependencyType.DEVENV);

    // THEN
    expect(() => p.deps.getDependency('zoo')).toThrow(/\"zoo\" is defined for multiple dependency types: runtime,devenv. Please specify dependency type/);
  });

  test('fails if type does not match', () => {
    // GIVEN
    const p = new TestProject();

    // WHEN
    p.deps.addDependency('zoo', DependencyType.RUNTIME);

    // THEN
    expect(() => p.deps.getDependency('zoo', DependencyType.BUILD)).toThrow(/there is no build dependency defined on \"zoo\"/);
  });

});

function depsManifest(p: Project) {
  p.synth();
  return JSON.parse(readFileSync(join(p.outdir, Dependencies.MANIFEST_FILE), 'utf-8'));
}