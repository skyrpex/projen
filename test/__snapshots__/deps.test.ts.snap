// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deps.all returns all the dependencies 1`] = `
Array [
  Object {
    "name": "dep1",
    "type": "devenv",
  },
  Object {
    "name": "dep3",
    "type": "peer",
  },
  Object {
    "name": "dep2",
    "type": "runtime",
  },
]
`;

exports[`minimal case 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "dependencies": Array [
    Object {
      "name": "my-first-dep",
      "type": "runtime",
    },
  ],
}
`;

exports[`with version requirement 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "dependencies": Array [
    Object {
      "name": "depy",
      "type": "peer",
      "version": "^7",
    },
  ],
}
`;
